server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -618,6 +618,65 @@ public Analyzer searchQuoteAnalyzer() {
         return this.searchQuoteAnalyzer;
     }
 
+    /**
+     * Builds a {@linkplain Function} to lookup mappers for a request, adding
+     * any {@code extraMapping} provided.
+     * @param runtimeMappings extra mappings parse and to add to the request
+     *        lookup or {@code null} if there aren't any extra mappings  
+     */
+    public Function<String, MappedFieldType> newFieldTypeLookup(Map<String, Object> runtimeMappings) {
+        if (runtimeMappings == null || runtimeMappings.size() == 0) {
+            return this::fieldType;
+        }
+        Mapper.BuilderContext builderContext = new Mapper.BuilderContext(indexSettings.getSettings(), new ContentPath(0));
+        Collection<ObjectMapper> objectMappers = new ArrayList<>();
+        Collection<FieldMapper> fieldMappers = new ArrayList<>();
+        Collection<FieldAliasMapper> fieldAliasMappers = new ArrayList<>();
+        for (Map.Entry<String, Object> runtimeEntry : runtimeMappings.entrySet()) {
+            @SuppressWarnings("unchecked") // Safe because that is how we deserialized it
+            Map<String, Object> definition = (Map<String, Object>) runtimeEntry.getValue();
+            String type = (String) definition.remove("type");
+            if (type == null) {
+                throw new IllegalArgumentException("[type] is required for runtime mapping [" + runtimeEntry.getKey() + "]");
+            }
+            Mapper.TypeParser parser = documentMapperParser().parserContext().typeParser(type);
+            if (parser == null) {
+                throw new IllegalArgumentException("[" + type + "] is unknown type for runtime mapping [" + runtimeEntry.getKey() + "]");
+            }
+            Mapper.Builder<?> builder = parser.parse(runtimeEntry.getKey(), definition, documentMapperParser().parserContext());
+            Mapper mapper = builder.build(builderContext);
+            
+            MapperUtils.collect(mapper, objectMappers, fieldMappers, fieldAliasMappers);
```

Comment:
Can you expand on why we need this? I was expecting that alias or objects would be caught anyways by your checks below, that only runtime fields can be defined.